Parallelize Dataset cleaner

ydailinkedin opened the pull request:
@ibuenros Can you review?

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -53,6 +71,10 @@ public DatasetCleaner(FileSystem fs, Properties props) throws IOException {
     }
+    this.executor =
+        Executors
+            .newFixedThreadPool(Integer.parseInt(props.getProperty(DATASET_CLEAN_THREAD_POOL_SIZE,
```
Ws should be using a cached thread pool instead of a fixed one and ideally avoid specifying a thread pool size as this is yet another knob that is hard to tune.

Instead, we should have an option that limits the rate of HDFS operations we perform. Then we should use as many threads are necessary to maintain that rate. @liyinan926 can give you some context as he already has added standard rate limiter implementations.

liyinan926 replied:
@chavdar using a cached thread pool poses the risk of having an unbounded number of threads running in the pool, which may happen if there are so many datasets to cleanup in a batch. This post talks about it and I think it makes sense http://dev.bizo.com/2014/06/cached-thread-pool-considered-harmlful.html.

liyinan926 replied:
I think it makes more sense to use a bounded variable size thread pool combined with a `RateBasedLimiter`.

chavdar replied:
Using a ThreadPoolExecutor with a high (say, 1000) thread limit and a core pool size of 0 should be fine. The point is we should not be using # of threads to control performance.

liyinan926 replied:
Yes, agreed.

ydailinkedin replied:
Understood. I will address the comments.

ibuenros commented:
LGTM other than Chavdar's comment.

ydailinkedin commented:
@chavdar @liyinan926 Updated.

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -12,28 +12,56 @@
 
   public static final String CONFIGURATION_KEY_PREFIX = "gobblin.retention.";
   public static final String DATASET_PROFILE_CLASS_KEY = CONFIGURATION_KEY_PREFIX + "dataset.profile.class";
+  public static final String DATASET_CLEAN_THREAD_POOL_SIZE = CONFIGURATION_KEY_PREFIX
```
I think max.concurrent.hdfs.calls is more descriptive.

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -12,28 +12,56 @@
   public static final String DATASET_PROFILE_CLASS_KEY = CONFIGURATION_KEY_PREFIX + "dataset.profile.class";
+  public static final String DATASET_CLEAN_THREAD_POOL_SIZE = CONFIGURATION_KEY_PREFIX
+      + "dataset.clean.thread.pool.size";
+  public static final String HDFS_CALL_RATE_LIMIT = CONFIGURATION_KEY_PREFIX + "hdfs.call.rate.limit";
```
Please be consistent in naming and use the dataset.clean prefix.

chavdar replied:
I think hdfs.calls.per.second.limit is more descriptive.

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -53,6 +81,13 @@ public DatasetCleaner(FileSystem fs, Properties props) throws IOException {
+        new ThreadPoolExecutor(0, Integer.parseInt(props.getProperty(DATASET_CLEAN_THREAD_POOL_SIZE,
+            DEFAULT_DATASET_CLEAN_THREAD_POOL_SIZE)), 0, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>());
+    this.service = MoreExecutors.listeningDecorator(executor);
+    if (StringUtils.isNotBlank(props.getProperty(HDFS_CALL_RATE_LIMIT))) {
```
We should have a default value for this, say 10.

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -61,9 +96,42 @@ public DatasetCleaner(FileSystem fs, Properties props) throws IOException {
+      Futures.addCallback(future, new FutureCallback<Void>() {
+        @Override
+        public void onFailure(Throwable arg0) {
+          finishCleanSignal.countDown();
```
Both failures and successes should ideally be exposed through a metric.

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetRetentionRateLimiter.java:
```diff
@@ -0,0 +1,71 @@
+import gobblin.runtime.RateBasedLimiter;
+
+/***
+ * Singleton class for controlling the hdfs call rate during datasets cleaning.
```
Why is this a singleton?

liyinan926 replied:
I think you want to do global throttling for the entire JVM instance. If this is necessary, you can try using Guice https://github.com/google/guice/wiki/Scopes instead of explicitly dealing with singleton.

ydailinkedin replied:
@chavdar I used a singleton since the rate control is for the total hdfs calls among all threads.

ibuenros replied:
If multiple filesystems are in use, it doesn't make sense to use a single rate limiter for all of them, I would suggest a registry of rate limiters keyed by URI.

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetRetentionRateLimiter.java:
```diff
@@ -0,0 +1,71 @@
+    return rateLimit;
+  }
+  
+  public static synchronized void setRateLimit(long rate) {
```
I don't think the rate limit should be mutable.

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetRetentionRateLimiter.java:
```diff
@@ -0,0 +1,71 @@
+    rateLimit = rate;
+  }
+  
+  public Limiter getRateLimiter() {
```
I don't see this used anywhere.

liyinan926 commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -61,9 +96,42 @@ public DatasetCleaner(FileSystem fs, Properties props) throws IOException {
+  @Override
+  public void close() throws IOException {
+    try {
+      finishCleanSignal.await();
```
Check for `null` before calling `await`.

liyinan926 commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -61,9 +96,42 @@ public DatasetCleaner(FileSystem fs, Properties props) throws IOException {
+    try {
+      finishCleanSignal.await();
+    } catch (InterruptedException e) {
+      throw new IOException("Not all datasets finish cleanning", e);
```
Do `Thread.currentThread().interrupt()` before the throw.

liyinan926 commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetRetentionRateLimiter.java:
```diff
@@ -0,0 +1,71 @@
+    return this.rateLimiter;
+  }
+
+  public void acquirePermits(long permits) {
```
You can create a convenience method `public void acquirePermit()` for acquiring a single permit since it's used a lot.

liyinan926 commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -61,9 +96,42 @@ public DatasetCleaner(FileSystem fs, Properties props) throws IOException {
+        @Override
+        public void onFailure(Throwable arg0) {
+          finishCleanSignal.countDown();
+          LOG.error("Failed to clean: " + dataset.datasetRoot());
```
Include the `Throwable` in the log message.

liyinan926 commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -12,28 +12,56 @@
 
   private final DatasetFinder datasetFinder;
+  private final ListeningExecutorService service;
+  private final List<ListenableFuture<?>> futures = Lists.newArrayList();
```
You don't need this list.

ibuenros commented:
Instead of adding getPermits to all calls in the code, you could implement a `ThrottledFileSystem`, which is just a decorator to `FileSystem`, and automatically gets permits for each call, that way the throttling is transparent to the developer. Care would be needed when creating derivated filesystems (like proxied file systems) to respect the throttling of the parent.

ydailinkedin commented:
@chavdar @liyinan926 @ibuenros  Updated. 

Also made changes in FileLevelTimestampVersionFinder. The previous logic is that a dataset version can only be found if there is a file existed along the dataset path. This could cause deleteEmptyDirectory being skipped. With the new patch, if no file is found, the empty directory itself will be treated as one dataset version, so that empty directories as well as its parent directories will be deleted.

ibuenros commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/RateControlledFileSystem.java:
```diff
@@ -0,0 +1,179 @@
+  }
+
+  @Override
+  public boolean delete(Path path) throws IOException {
```
This calls delete(Path, boolean), so no need to acquire permits here.

ydailinkedin replied:
Has to implement it since `delete(Path)` is abstract in `FileSystem`.

ibuenros commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/RateControlledFileSystem.java:
```diff
@@ -0,0 +1,179 @@
+  }
+
+  @Override
+  public boolean mkdirs(Path path) throws IOException {
```
This calls mkdirs(Path, FsPermission)

ibuenros commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/RateControlledFileSystem.java:
```diff
@@ -0,0 +1,179 @@
+    this.rateLimiter.stop();
+  }
+
+  private void acquirePermit() throws IOException {
```
I would rather not store the rateLimiter as an instance variable, instead getting it from the cache every time acquirePermit() is called. This way, if the rateLimiter ever gets cleaned from cache, we will not end up with a new file system in the same uri using a different rate limiter.

ibuenros commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/RateControlledFileSystem.java:
```diff
@@ -0,0 +1,179 @@
+ *  to make sure HDFS call rate is allowed by the uppper limit.
+ *  </p>
+ */
+public class RateControlledFileSystem extends FileSystem {
```
This class should implement Decorator interface.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/FileListUtils.java:
```diff
@@ -54,6 +55,45 @@ public boolean accept(Path path) {
+      if (ArrayUtils.isEmpty(fs.listStatus(fileStatus.getPath()))) {
+        files.add(fileStatus);
+      } else {
+        for (FileStatus status : fs.listStatus(fileStatus.getPath())) {
```
Can you reuse fs.listStatus? You're calling it twice in a row.

ibuenros commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -12,36 +12,81 @@
 
   public static final String CONFIGURATION_KEY_PREFIX = "gobblin.retention.";
   public static final String DATASET_PROFILE_CLASS_KEY = CONFIGURATION_KEY_PREFIX + "dataset.profile.class";
+  public static final String DATASET_CLEAN_MAX_CONCURRENT_HDFS_CALLS = CONFIGURATION_KEY_PREFIX
```
Can we rename this to max_concurrent_datasets_cleaned instead?

ibuenros commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -12,36 +12,81 @@
+  public static final String DATASET_CLEAN_HDFS_CALLS_PER_SECOND_LIMIT = CONFIGURATION_KEY_PREFIX + "hdfs.calls.per.second.limit";
+
+  public static final String DEFAULT_DATASET_CLEAN_MAX_CONCURRENT_HDFS_CALLS = "1000";
+  public static final String DEFAULT_DATASET_CLEAN_HDFS_CALLS_PER_SECOND_LIMIT = "10";
```
This default seems low. I would rather make it infinite, or at least very large (most users will not change this, and it would suck if the cleaner is slow by default).

ydailinkedin commented:
@ibuenros Updated.

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -52,7 +102,18 @@ public DatasetCleaner(FileSystem fs, Properties props) throws IOException {
+            DEFAULT_MAX_CONCURRENT_DATASETS_CLEANED)), 0, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<Runnable>());
+    this.service = MoreExecutors.listeningDecorator(executor);
+
+    this.metricContext = this.closer.register(new MetricContext.Builder(DatasetCleaner.class.getName()).build());
```
There should be an optional parent metric context.

chavdar commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/RateControlledFileSystem.java:
```diff
@@ -0,0 +1,195 @@
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.retention;
```
I think this is a pretty useful class even outside this context. I think gobblin-utlity might be a better place for it.

chavdar commented:
Please add some unit tests.

liyinan926 commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleaner.java:
```diff
@@ -61,9 +122,93 @@ public DatasetCleaner(FileSystem fs, Properties props) throws IOException {
+  @Override
+  public void close() throws IOException {
+    try {
+      if (this.finishCleanSignal != null) {
```
You can make `finishCleanSignal` an `Optional`.

liyinan926 commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/DatasetCleanerJob.java:
```diff
@@ -47,16 +49,21 @@ public DatasetCleanerJob(String id, Props props) throws IOException {
-      this.datasetCleaner.clean();
+  public void run() throws Exception {
+    if (this.datasetCleaner != null) {
+      Closer closer = Closer.create();
```
Seems this `closer` is not used.

liyinan926 commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/RateControlledFileSystem.java:
```diff
@@ -0,0 +1,195 @@
+ *  to make sure HDFS call rate is allowed by the uppper limit.
+ *  </p>
+ */
+public class RateControlledFileSystem extends FileSystem implements Closeable, Decorator {
```
Seems `Decorator` should be parameterized with a type parameter.

liyinan926 replied:
`FileSystem` already implements `Closeable`.

liyinan926 commented on gobblin-data-management/src/main/java/gobblin/data/management/retention/RateControlledFileSystem.java:
```diff
@@ -0,0 +1,195 @@
+public class RateControlledFileSystem extends FileSystem implements Closeable, Decorator {
+
+  private static final int DEFAULT_MAX_CACHE_SIZE = 100;
+  private static final Cache<String, Limiter> FS_URI_TO_RATE_LIMITER_CACHE = CacheBuilder.newBuilder()
```
It's better to use a `CacheLoader` and initialize this in the constructor instead of creating a `Callable` every time `get` is called in `getRateLimiter`.

liyinan926 commented on gobblin-utility/src/main/java/gobblin/util/FileListUtils.java:
```diff
@@ -54,6 +55,46 @@ public boolean accept(Path path) {
+   */
+  public static List<FileStatus> listMostNestedPathRecursively(FileSystem fs, Path path) throws FileNotFoundException,
+      IOException {
+    return listMostNestedPathRecursively(fs, path, new PathFilter() {
```
It's worth defining this special `PathFilter` as a constant like `private static final PathFilter NO_OP_PATH_FILTER = ...`.

ydailinkedin commented:
@chavdar @liyinan926 Updated. 
Moved `RateControlledFileSystem` to `gobblin-runtime`. I tried to move it to `gobblin-utility`, but turns out it would cause dependency loop. Also, moving all `limiter` related classes to `gobblin-utility` may not be a good idea since it involves api changes.

Addressed rest comments.

ibuenros commented:
Talked with @ydailinkedin already.
I don't think gobblin-data-management should depend on gobblin-runtime, as most of gobblin-runtime and gobblin-core is irrelevant to data management (apparently everything except the RateLimiters).
RateLimiter should move to gobblin-utility. While this is in some way an API change, it would be a very weird setup where someone has gobblin-runtime in the classpath but not gobblin-utility, and in particular Gobblin ETL depends on gobblin-utility (I think), so it most likely would not break anything.

chavdar commented:
@ydailinkedin @ibuenros @liyinan926 I agree that gobblin-utility is a better home for the rate limitter than gobblin-runtime. However moving to gobblin-utility should entail changing the namespace which is a backwards-incompatible API change, otherwise it's going to be confusing. A compromise option can be to (a) move the rate-limitter classes to gobblin-utlity/gobblin.util, retain the existing runtime classes as deprecated and make them extend the gobblin.util classes.

Let's discuss this F2F.

ydailinkedin commented:
@chavdar @liyinan926 @ibuenros Updated. Removed the dependency on gobblin-runtime from gobblin-data-management.

ibuenros commented:
LGTM
